src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,500 @@
+            var updatedSemanticDocument = await SemanticDocument.CreateAsync(updatedSolutionWithParameter.GetDocument(document.Document.Id), cancellationToken).ConfigureAwait(false);
+
+            var newExpression = (TExpressionSyntax)updatedSemanticDocument.Root.GetAnnotatedNodesAndTokens(annotatedExpression).Single().AsNode()!;
+            var documentWithUpdatedMethodBody = ConvertExpressionWithNewParameter(updatedSemanticDocument, newExpression, parameterName, allOccurrences, cancellationToken);
+            return documentWithUpdatedMethodBody.Project.Solution;
+        }
+
+        protected static Task<Solution> AddParameterToMethodHeaderAsync(SemanticDocument document, TExpressionSyntax expression, string parameterName, CancellationToken cancellationToken)
+        {
+            var invocationDocument = document.Document;
+            var syntaxFacts = invocationDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var methodExpression = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.isMethodDeclaration(node), ascendOutOfTrivia: true)!;
+            var semanticModel = document.SemanticModel;
+            var symbolInfo = (IMethodSymbol)semanticModel.GetRequiredDeclaredSymbol(methodExpression, cancellationToken);
+            var parameterType = semanticModel.GetTypeInfo(expression, cancellationToken).Type ?? document.SemanticModel.Compilation.ObjectType;
+            var refKind = syntaxFacts.GetRefKindOfArgument(expression);
+
+            return AddParameterService.Instance.AddParameterAsync(
+                invocationDocument,
+                symbolInfo,
+                parameterType,
+                refKind,
+                parameterName,
+                newParameterIndex: null,
+                fixAllReferences: false,
+                cancellationToken);
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var invocationDocument = document.Document;
+            var semanticModel = document.SemanticModel;
+            var semanticFacts = invocationDocument.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(
+                    semanticModel, expression, capitalize: false, cancellationToken: cancellationToken);
+        }
+
+        protected static IMethodSymbol GetMethodSymbolFromExpression(SemanticDocument annotatedSemanticDocument, TExpressionSyntax annotatedExpression, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = annotatedSemanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var methodExpression = annotatedExpression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.isMethodDeclaration(node), true)!;
```

Comment:
1. named parameter for 'self'. 
2. why the `!`.  how do you know this will always succeed?

Reply:
At this point we should know that we are contained within some sort of method since the code action does not appear if that is false.

Reply:
Oh I looked at this again and realized I did use this in the function that determines whether or not code actions should appear, so I changed it.